### PR TITLE
Refactor reset PIN flow out of UserSession

### PR DESCRIPTION
**Pull request by vinaysshenoy**
https://www.pivotaltracker.com/story/show/168980786

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/user/resetpin/ResetUserPin.kt [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        .flatMap { loginApi.resetPin(it) }
+        .doOnSubscribe { Timber.i("Resetting PIN") }
+
+    val updateUserOnSuccess = resetPasswordCall
+        .flatMapCompletable { storeUserAndAccessToken(it) }
+        .toSingleDefault(ResetPinResult.Success as ResetPinResult)
+
+    return updateUserOnSuccess
+        .onErrorReturn {
+          when (it) {
```
Comment (ragunathjawahar) [inline app/src/main/java/org/simple/clinic/user/resetpin/ResetUserPin.kt]: Extract this block to a named function?

File: app/src/test/java/org/simple/clinic/forgotpin/confirmpin/ForgotPinConfirmPinScreenControllerTest.kt [hunk 2]
```diff
@@ -64,8 +66,8 @@ class ForgotPinConfirmPinScreenControllerTest {
 
   @Before
   fun setUp() {
-    whenever(userSession.requireLoggedInUser()) doReturn Observable.just(loggedInUser)
-    whenever(facilityRepository.currentFacility(any<User>())) doReturn Observable.just(facility)
```
Comment (ragunathjawahar) [inline app/src/test/java/org/simple/clinic/forgotpin/confirmpin/ForgotPinConfirmPinScreenControllerTest.kt]: 👍

File: app/src/test/java/org/simple/clinic/forgotpin/confirmpin/ForgotPinConfirmPinScreenControllerTest.kt [hunk 3]
```diff
@@ -93,32 +90,22 @@ class ForgotPinConfirmPinScreenControllerTest {
   }
 
   @Test
-  @Parameters(value = [
-    "0000|000|0000|false",
-    "0000|000|001|true",
-    "0000|0000|000|false"
-  ])
```
Comment (ragunathjawahar) [inline app/src/test/java/org/simple/clinic/forgotpin/confirmpin/ForgotPinConfirmPinScreenControllerTest.kt]: 👌

File: app/src/main/java/org/simple/clinic/user/resetpin/ResetUserPin.kt (hunk 1 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/user/resetpin/ResetUserPin.kt]: This was done in a later commit.

File: app/src/test/java/org/simple/clinic/user/resetpin/JavaHashPasswordHasher.kt [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+class JavaHashPasswordHasher : PasswordHasher {
+
+  override fun hash(password: String): Single<String> {
+    return Single.just(genHash(password))
+  }
+
+  override fun compare(hashed: String, password: String): Single<ComparisonResult> {
+    val hashedPassword = genHash(password)
+
+    val result = if (hashed == hashedPassword) SAME else DIFFERENT
```
Comment (ragunathjawahar) [inline app/src/test/java/org/simple/clinic/user/resetpin/JavaHashPasswordHasher.kt]: Nice!

File: app/src/test/java/org/simple/clinic/user/resetpin/ResetUserPinTest.kt [hunk 5]
```diff
@@ -39,19 +38,21 @@ import java.util.UUID
 @RunWith(JUnitParamsRunner::class)
 class ResetUserPinTest {
 
-  private val passwordHasher = mock<PasswordHasher>()
+  private val passwordHasher = JavaHashPasswordHasher()
```
Comment (ragunathjawahar) [inline app/src/test/java/org/simple/clinic/user/resetpin/ResetUserPinTest.kt]: 👍

**Pull request merged**